<a href="https://colab.research.google.com/github/prerit-spec/Pyhton-Data-Analysis/blob/main/KNN_vs_Naive_Bayes_(Spam_Detection).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Naive Bayes Classifier vs KNN (Spam detection)
### Cse minor (AIML) Teacher Assesment
#### Name: Prerit Rudrakar
#### Electronics Department
#### Roll no: A-55


Here in this notebook I am going to compare the accuracy of 'Naive Bayes Classifiers' and 'KNN' on a Spam Detection Dataset.

In [1]:
# Importing all the required libraries.
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


In [2]:
# Reading the dataset.
dataset = pd.read_csv('/spam_ham_dataset.csv')

In [3]:
# Determining the columns available.
dataset.columns

Index(['Unnamed: 0', 'label', 'text', 'label_num'], dtype='object')

In [4]:
#Printing the dataset first five rows.
dataset.head(5)

Unnamed: 0  ... label_num
0         605  ...         0
1        2349  ...         0
2        3624  ...         0
3        4685  ...         1
4        2030  ...         0

[5 rows x 4 columns]

In [5]:
#Determining the size of the dataset.
dataset.shape

(5171, 4)

In [6]:
dataset["label_num"].value_counts

<bound method IndexOpsMixin.value_counts of 0       0
1       0
2       0
3       1
4       0
       ..
5166    0
5167    0
5168    0
5169    0
5170    1
Name: label_num, Length: 5171, dtype: int64>

In [7]:
dataset['Unnamed: 0'].value_counts

<bound method IndexOpsMixin.value_counts of 0        605
1       2349
2       3624
3       4685
4       2030
        ... 
5166    1518
5167     404
5168    2933
5169    1409
5170    4807
Name: Unnamed: 0, Length: 5171, dtype: int64>

Since we cannot use Unnamed for classification we are dropping it

In [8]:
dataset.drop('Unnamed: 0',axis=1,inplace=True)

We can have either label or label_num 

By observing the top 5 datapoints in dataset

if label is ham; label_num is 0 

if label is spam; label_num is 1

In [9]:
dataset.drop('label',axis=1,inplace=True)

In [10]:
dataset.head(5)

text  label_num
0  Subject: enron methanol ; meter # : 988291\r\n...          0
1  Subject: hpl nom for january 9 , 2001\r\n( see...          0
2  Subject: neon retreat\r\nho ho ho , we ' re ar...          0
3  Subject: photoshop , windows , office . cheap ...          1
4  Subject: re : indian springs\r\nthis deal is t...          0

Converting text to

In [11]:
#preprocessign the text.
import re
def preprocess(text):
    text = text.replace('\r',' ')
    text = text.replace('\n',' ')
    text = text.replace('#',' ')
    text = text.replace("we ' re","we are")
    text = text.replace("they ' re","they are")
    text = text.replace("you ' re","you are")
    text = text.replace("Subject:"," ")

    return text

In [12]:
dataset['text_clean']=dataset['text'].map(preprocess)

In [13]:
dataset.head(5)

text  ...                                         text_clean
0  Subject: enron methanol ; meter # : 988291\r\n...  ...    enron methanol ; meter   : 988291  this is a...
1  Subject: hpl nom for january 9 , 2001\r\n( see...  ...    hpl nom for january 9 , 2001  ( see attached...
2  Subject: neon retreat\r\nho ho ho , we ' re ar...  ...    neon retreat  ho ho ho , we are around to th...
3  Subject: photoshop , windows , office . cheap ...  ...    photoshop , windows , office . cheap . main ...
4  Subject: re : indian springs\r\nthis deal is t...  ...    re : indian springs  this deal is to book th...

[5 rows x 3 columns]

Converting text_clean to vector

In [14]:
from sklearn.feature_extraction.text import CountVectorizer


count_vectorizer = CountVectorizer(ngram_range=(1,1))
X = count_vectorizer.fit_transform(dataset['text_clean'])
print(X.shape)

(5171, 50447)


Splitting the dataset into train ans test dataset

In [15]:
from sklearn.model_selection import train_test_split

x,x_test,y,y_test = train_test_split(X,dataset['label_num'],test_size=0.2,random_state=42)
x_train,x_cv,y_train,y_cv = train_test_split(x,y,test_size=0.2,random_state=42)

In [16]:
print((x_train).shape)

(3308, 50447)


In [17]:
import numpy as np
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
error=[]
from sklearn.neighbors import KNeighborsClassifier
for i in range(1,20):
    neigh = KNeighborsClassifier(n_neighbors=i)
    neigh.fit(x_train,y_train)
    y_pred = neigh.predict(x_cv)
    print(classification_report(y_cv, y_pred))
    print(confusion_matrix(y_cv, y_pred))
    print()
    print()
    print(accuracy_score(y_cv, y_pred))
    print("*"*100)
    
    

              precision    recall  f1-score   support

           0       0.95      0.90      0.93       587
           1       0.79      0.88      0.83       241

    accuracy                           0.90       828
   macro avg       0.87      0.89      0.88       828
weighted avg       0.90      0.90      0.90       828

[[531  56]
 [ 30 211]]


0.8961352657004831
****************************************************************************************************
              precision    recall  f1-score   support

           0       0.91      0.92      0.91       587
           1       0.80      0.78      0.79       241

    accuracy                           0.88       828
   macro avg       0.85      0.85      0.85       828
weighted avg       0.88      0.88      0.88       828

[[540  47]
 [ 54 187]]


0.8780193236714976
****************************************************************************************************
              precision    recall  f1-score   support

 

First we will use K Neighbor classifiers with K=1.

In [18]:
neigh = KNeighborsClassifier(n_neighbors=1)
neigh.fit(x_train,y_train)
y_pred = neigh.predict(x_test)
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.89      0.92       742
           1       0.76      0.85      0.81       293

    accuracy                           0.88      1035
   macro avg       0.85      0.87      0.86      1035
weighted avg       0.89      0.88      0.88      1035

[[664  78]
 [ 43 250]]
0.8830917874396135


We have got accuracy of 88% using KNN with K=1.

Now we'll use Gaussian naive bayes classifiers.

In [19]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()
y_pred = gnb.fit(x.toarray(), y).predict(x_test.toarray())
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(accuracy_score(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.95      0.99      0.97       742
           1       0.96      0.87      0.92       293

    accuracy                           0.95      1035
   macro avg       0.96      0.93      0.94      1035
weighted avg       0.95      0.95      0.95      1035

[[732  10]
 [ 37 256]]
0.9545893719806763


 When using naive bayes we get an accuracy of 95.4% f1 score =0.97

## So we may concluse that the Naive Bayes algorithm  is more accurate than K Nearest Neighbor.

##### Naive Bayes is a linear classifier while K-NN is not.
#####It tends to be faster when applied to big data. In comparison, k-nn is usually slower for large amounts of data, because of the calculations required for each new step in the process. If speed is important, choose Naive Bayes over K-NN.